universal-application-tool-0.0.1/app/services/applicant/ApplicantData.java

```diff
@@ -133,8 +136,59 @@ public boolean equals(@Nullable Object object) {
     return false;
   }
 
+  public boolean hasPath(Path path) {
+    try {
+      return read(path, Object.class).isPresent();
+    } catch (JsonPathTypeMismatchException e) {
+      return false;
+    }
+  }
+
   @Override
   public int hashCode() {
     return Objects.hash(jsonData.jsonString());
   }
+
+  /**
+   * Copies all keys from {@code other}, recursively. All lists are merged. No values will be
+   * overwritten.
+   *
+   * @return A list of {@code Path}s whose values could not be copied due to conflicts.
+   */
+  public ImmutableList<Path> mergeFrom(ApplicantData other) {
+    Map<?, ?> rootAsMap = other.jsonData.read("$", Map.class);
+    return mergeFrom(Path.empty(), rootAsMap);
+  }
+
+  private ImmutableList<Path> mergeFrom(Path rootKey, Map<?, ?> other) {
+    ImmutableList.Builder<Path> pathsRemoved = new ImmutableList.Builder<>();
+    for (Map.Entry<?, ?> entry : other.entrySet()) {
+      String key = entry.getKey().toString();
+      Path path = rootKey.toBuilder().append(key).build();
+      if (hasPath(path)) {
+        if (entry.getValue() instanceof Map) {
+          // Recurse into maps.
+          pathsRemoved.addAll(mergeFrom(path, (Map) entry.getValue()));
+        } else if (entry.getValue() instanceof List) {
+          // Add items from lists.
+          for (Object item : (List) entry.getValue()) {
```

Comment:
(non-blocking) Can you add a `TODO` here to fix this once we've figured out how repeated data will be stored? You could mention #405 or we could open a more specific one.

One thing that might help is to pull it into a method like `mergeList` with the TODO just to keep the complexity nice and isolated.

Reply:
Sure.  I think this is a reasonable stub for now but I agree it won't work for lists of objects, if we have those.  Added reference to #405.  Thanks!